lib/battle/BattleHexArray.h

```diff
@@ -0,0 +1,321 @@
+/*
+ * BattleHexArray.h, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#pragma once
+
+#include "BattleHex.h"
+
+VCMI_LIB_NAMESPACE_BEGIN
+
+/// Class representing an array of unique BattleHex objects
+class DLL_LINKAGE BattleHexArray
+{
+public:
+	static constexpr uint8_t totalSize = GameConstants::BFIELD_SIZE;
+	using StorageType = std::vector<BattleHex>;
+
+	using value_type = BattleHex;
+	using size_type = StorageType::size_type;
+	using reference = value_type &;
+	using const_reference = const value_type &;
+	using pointer = value_type *;
+	using const_pointer = const value_type *;
+	using difference_type = typename StorageType::difference_type;
+	using iterator = typename StorageType::iterator;
+	using const_iterator = typename StorageType::const_iterator;
+	using reverse_iterator = typename StorageType::reverse_iterator;
+	using const_reverse_iterator = typename StorageType::const_reverse_iterator;
+
+	using ArrayOfBattleHexArrays = std::array<BattleHexArray, GameConstants::BFIELD_SIZE>;
+
+	static const ArrayOfBattleHexArrays neighbouringTilesCache;
+	static const std::map<BattleSide, ArrayOfBattleHexArrays> neighbouringTilesDblWide;
```

Comment:
well... `neighbouringTilesCache` was its original naming. But of course you are right. These are precomputed values. I firstly done them in class scope, then in function scope and currently they are in class scope again (wanted to see if there is performance gain in using those statics in class scope), however I think there might be the risk of ["static initialization order fiasco"](https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use-members) and probably it would be best to initialize them in function scope as I did before.